chain/update.go

```diff
@@ -0,0 +1,150 @@
+package chain
+
+import (
+	"time"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils/chain"
+	"go.sia.tech/coreutils/wallet"
+	"go.sia.tech/renterd/api"
+)
+
+type (
+	Update struct {
+		Index types.ChainIndex
+
+		ContractUpdates     map[types.FileContractID]*ContractUpdate
+		HostUpdates         map[types.PublicKey]HostUpdate
+		WalletOutputUpdates map[types.Hash256]WalletOutputUpdate
+		WalletEventUpdates  []WalletEventUpdate
+	}
+
+	ContractUpdate struct {
+		ProofHeight *uint64
+		Size        *uint64
+		State       api.ContractState
+
+		RevisionHeight *uint64
+		RevisionNumber *uint64
+	}
+
+	HostUpdate struct {
+		Announcement chain.HostAnnouncement
+		BlockHeight  uint64
+		BlockID      types.BlockID
+		Timestamp    time.Time
+	}
+
+	WalletEventUpdate struct {
+		Addition bool
+		Event    wallet.Event
+	}
+
+	WalletOutputUpdate struct {
+		Addition bool
+		Element  wallet.SiacoinElement
+		ID       types.Hash256
+	}
+)
+
+// NewChainUpdate returns a new ChainUpdate.
+func NewChainUpdate() *Update {
+	return &Update{
+		ContractUpdates:     make(map[types.FileContractID]*ContractUpdate),
+		HostUpdates:         make(map[types.PublicKey]HostUpdate),
+		WalletOutputUpdates: make(map[types.Hash256]WalletOutputUpdate),
+	}
+}
+
+// ContractUpdate returns the ContractUpdate for the given file contract ID. If
+// it doesn't exist we initialize one with the given state.
+func (cu *Update) ContractUpdate(fcid types.FileContractID, state api.ContractState) *ContractUpdate {
+	_, ok := cu.ContractUpdates[fcid]
+	if !ok {
+		cu.ContractUpdates[fcid] = &ContractUpdate{State: state}
+	}
+	return cu.ContractUpdates[fcid]
+}
+
+// AddEvents is called with all relevant events added in the update.
+func (cu *Update) AddEvents(events []wallet.Event) error {
+	for _, event := range events {
+		cu.WalletEventUpdates = append(cu.WalletEventUpdates, WalletEventUpdate{
+			Addition: true,
+			Event:    event,
+		})
+	}
+	return nil
+}
+
+// AddSiacoinElements is called with all new siacoin elements in the
+// update. Ephemeral siacoin elements are not included.
+func (cu *Update) AddSiacoinElements(ses []wallet.SiacoinElement) error {
+	for _, se := range ses {
+		cu.WalletOutputUpdates[se.ID] = WalletOutputUpdate{
+			Addition: true,
+			ID:       se.ID,
+			Element:  se,
+		}
+	}
+	return nil
+}
+
+// RemoveSiacoinElements is called with all siacoin elements that were
+// spent in the update.
+func (cu *Update) RemoveSiacoinElements(ids []types.SiacoinOutputID) error {
+	for _, id := range ids {
+		cu.WalletOutputUpdates[types.Hash256(id)] = WalletOutputUpdate{
+			Addition: false,
+			ID:       types.Hash256(id),
+		}
+	}
+	return nil
+}
+
+// WalletStateElements returns all state elements in the database. It is used
+// to update the proofs of all state elements affected by the update.
+func (cu *Update) WalletStateElements() (elements []types.StateElement, _ error) {
```

Comment:
Comment says "all elements in the database" but this only returns all elements within the current `Update`.
I'm actually kinda surprised nothing breaks. We should probably make sure testing catches that.

Reply:
I updated this, still have to update the test but wanted to get an OK in reviews first. Not too sure yet how to write a test for this.